nightfall-deployer/contracts/X509.sol

```diff
@@ -0,0 +1,345 @@
+        bytes memory message = messageTlv.octets;
+        return message;
+    }
+
+    function modExp(
+        bytes memory b,
+        uint256 e,
+        bytes memory m
+    ) internal view returns (bytes memory) {
+        bool success;
+        bytes memory result;
+        (success, result) = (
+            address(5).staticcall(abi.encodePacked(b.length, uint256(32), m.length, b, e, m))
+        );
+        require(success, 'modExp error');
+        return result;
+    }
+
+    /*
+    Validate the decrypted signature and returns the message hash
+    */
+    function validateSignatureAndExtractMessageHash(bytes memory decrypt, uint256 tlvLength)
+        private
+        view
+        returns (bytes memory)
+    {
+        DecodedTlv[] memory tlvs = new DecodedTlv[](tlvLength);
+        require(
+            decrypt[0] == 0x00 && decrypt[1] == 0x00,
+            'Decrypt does not have a leading zero octets'
+        );
+        require(
+            decrypt[2] == 0x00 || decrypt[2] == 0x01,
+            'Block Type is not a private key operation'
+        );
+        // loop through the padding
+        uint256 i;
+        for (i = 3; i < decrypt.length; i++) {
+            if (decrypt[i] != 0xff) break;
+        }
+        i++;
+        tlvs = this.parseDER(decrypt, i, tlvLength);
+        require(
+            tlvs[4].depth == 1 && tlvs[4].tag.tagType == 0x04,
+            'Incorrect tag or position for decrypted hash data'
+        );
+        bytes memory messageHashFromSignature = tlvs[4].value;
+
+        return messageHashFromSignature;
+    }
+
+    // note: this function is from an MIT licensed library, with appreciation to
+    // https://github.com/bokkypoobah/BokkyPooBahsDateTimeLibrary/blob/v1.01/contracts/BokkyPooBahsDateTimeLibrary.sol
+    // minor changes made
+    function timestampFromDate(bytes memory utcTime) private pure returns (uint256 _seconds) {
+        uint256 year = uint256(uint8(utcTime[0]) - 48) *
+            10 +
+            uint256(uint8(utcTime[1]) - 48) +
+            2000;
+        uint256 month = uint256(uint8(utcTime[2]) - 48) * 10 + uint256(uint8(utcTime[3]) - 48);
+        uint256 day = uint256(uint8(utcTime[4]) - 48) * 10 + uint256(uint8(utcTime[5]) - 48);
+        require(year >= 1970);
+        int256 _year = int256(year);
+        int256 _month = int256(month);
+        int256 _day = int256(day);
+
+        int256 __days = _day -
+            32075 +
+            (1461 * (_year + 4800 + (_month - 14) / 12)) /
+            4 +
+            (367 * (_month - 2 - ((_month - 14) / 12) * 12)) /
+            12 -
+            (3 * ((_year + 4900 + (_month - 14) / 12) / 100)) /
+            4 -
+            OFFSET19700101;
+
+        _seconds = uint256(__days) * SECONDS_PER_DAY;
+    }
+
+    // this function finds and checks the Not Before and Not After tlvs
+    // TODO this and the subsequent function loop over the tlvs twice - this is inefficient - refactor code
```

Comment:
I don't see a problem with the loop in these functions, the worst Time Complexity is `O(n)` and I might be wrong but the `tlvs` array doesn't seem to get too large, anyways inside the loop you don't do any intensive operation. I can't see much room for improvement honestly, then, if you agree, you can remove the TODO. :)

Reply:
I was thinking that the functions each perform a similar loop so maybe they could be combined in future.  I'll remove it though because it's not needed.